### PR TITLE
Split linker interface into separate artifact

### PR DESCRIPTION
**Pull request by gzm0**
This is the first step for #3661 / #3714.

### CODE REVIEW

**Comment by sjrd** [comment]
The new package structure looks good to me. :)

The `interface.unstable` package makes me wonder whether we should move all `linker.x` subpackages (except `interface` and `standard`) in a subpackage `linker.unstable.x`?

**Comment by gzm0** [comment]
> The interface.unstable package makes me wonder whether we should move all linker.x subpackages (except interface and standard) in a subpackage linker.unstable.x?

Yes, potentially.

**Comment by gzm0** [comment]
This is ready for final review. I think I'll put the actual reflective loading into a separate PR.

We might want to wait with merging until the reflective approach is validated.

File: linker/shared/src/main/scala/org/scalajs/linker/interface/unstable/OutputFileImpl.scala [hunk 1]
```diff
@@ -31,7 +30,17 @@ abstract class OutputFileImpl extends LinkerOutput.File {
         else Future.successful(())
       }
 
-      writeLoop().finallyWith(chan.close())
+      writeLoop()
+      // finallyWith(chan.close())
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/interface/unstable/OutputFileImpl.scala]: Consider extracting that in a
```scala
private def finallyWith[A](f: Future[A], finallyFuture: => Future[Unit]): Future[A]
```
instead of relying on a comment.

File: linker/js/src/main/scala/org/scalajs/linker/NodeIRContainer.scala [hunk 2]
```diff
@@ -23,15 +23,15 @@ import scala.scalajs.js.typedarray.TypedArrayBufferOps._
 
 import java.nio._
 
-import org.scalajs.linker.standard.{IRContainerImpl, IRFileImpl}
+import org.scalajs.linker.interface.{IRContainer, IRFile}
+import org.scalajs.linker.interface.unstable.IRContainerImpl
 import org.scalajs.linker.standard.MemIRFileImpl
 
-abstract class IRContainerPlatformExtensions private[linker] () {
+object NodeIRContainer {
```
Comment (sjrd) [inline linker/js/src/main/scala/org/scalajs/linker/NodeIRContainer.scala]: Shouldn't this refactoring be in its own commit, like the other refactorings in this PR?

File: Jenkinsfile [hunk 3]
```diff
@@ -308,20 +308,22 @@ def Tasks = [
         'set scalaJSStage in testSuite := FastOptStage' \
         ++$scala testSuite/bootstrap:test &&
     sbt ++$scala irJS/mimaReportBinaryIssues \
-        loggingJS/mimaReportBinaryIssues linkerJS/mimaReportBinaryIssues
+        loggingJS/mimaReportBinaryIssues \
+	linkerInterfaceJS/mimaReportBinaryIssues linkerJS/mimaReportBinaryIssues
```
Comment (sjrd) [inline Jenkinsfile]: There's a tab here, and in quite a few other places in this file.

File: build.sbt [hunk 4]
```diff
@@ -7,7 +7,9 @@ val compiler = Build.compiler
 val logging = Build.logging
 val loggingJS = Build.loggingJS
 val linker = Build.linker
+val linkerInterface = Build.linkerInterface
```
Comment (sjrd) [inline build.sbt]: For consistency, these declarations should follow the order of declarations in `project/Build.scala`.

**Comment by gzm0** [comment]
FYI, updates are still WIP.

**Comment by gzm0** [comment]
Updated. I think I have addressed everything.
